drivers/virtio/virtio-mmio.c

```diff
@@ -24,212 +24,852 @@
+ ****************************************************************************/
+
+static uint8_t virtio_mmio_get_status(FAR struct virtio_device *vdev)
+{
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  return metal_io_read32(&vmdev->cfg_io, VIRTIO_MMIO_STATUS);
+}
+
+/****************************************************************************
+ * Name: virtio_mmio_write_config
+ ****************************************************************************/
+
+static void virtio_mmio_write_config(FAR struct virtio_device *vdev,
+                                     uint32_t offset, void *src, int length)
+{
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  uint32_t write_offset = VIRTIO_MMIO_CONFIG + offset;
+  uint32_t u32data;
+  uint16_t u16data;
+  uint8_t u8data;
+
+  if (vdev->id.version == VIRTIO_MMIO_VERSION_1)
+    {
+      FAR char *s = src;
+      int i;
+      for (i = 0; i < length; i++)
+        {
+          metal_io_write8(&vmdev->cfg_io, write_offset + i, s[i]);
+        }
+
+      return;
+    }
+
+  switch (length)
+    {
+      case 1:
+        memcpy(&u8data, src, sizeof(u8data));
+        metal_io_write8(&vmdev->cfg_io, write_offset, u8data);
+        break;
+      case 2:
+        memcpy(&u16data, src, sizeof(u16data));
+        metal_io_write16(&vmdev->cfg_io, write_offset, u16data);
+        break;
+      case 4:
+        memcpy(&u32data, src, sizeof(u32data));
+        metal_io_write32(&vmdev->cfg_io, write_offset, u32data);
+        break;
+      case 8:
+        memcpy(&u32data, src, sizeof(u32data));
+        metal_io_write32(&vmdev->cfg_io, write_offset, u32data);
+        memcpy(&u32data, src + sizeof(u32data), sizeof(u32data));
+        metal_io_write32(&vmdev->cfg_io, write_offset + sizeof(u32data),
+                         u32data);
+        break;
+      default:
+        DEBUGASSERT(0);
+    }
+}
+
+/****************************************************************************
+ * Name: virtio_mmio_read_config
+ ****************************************************************************/
+
+static void virtio_mmio_read_config(FAR struct virtio_device *vdev,
+                                    uint32_t offset, FAR void *dst,
+                                    int length)
+{
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  uint32_t read_offset = VIRTIO_MMIO_CONFIG + offset;
+  uint32_t u32data;
+  uint16_t u16data;
+  uint8_t u8data;
+
+  if (vdev->id.version == VIRTIO_MMIO_VERSION_1)
+    {
+      FAR char *d = dst;
+      int i;
+      for (i = 0; i < length; i++)
+        {
+          d[i] = metal_io_read8(&vmdev->cfg_io, read_offset + i);
+        }
+
+      return;
+    }
 
-  virtq->desc[id].addr = (uintptr_t)addr;
-  virtq->desc_virt[id] = addr;
-  return id;
+  switch (length)
+    {
+      case 1:
+        u8data = metal_io_read8(&vmdev->cfg_io, read_offset);
```

Comment:
memcpy(dst, metal_io_read8(&vmdev->cfg_io, read_offset), 1);

Reply:
metal_io_read8(&vmdev->cfg_io, read_offset) return a value rather than a pointer, so we can change like this.